Fix Elasticsearch exceptions

steveburkett opened the pull request:
https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/392

steveburkett committed:
Use the Aws::InstanceProfileCredentials object for aws creds

Since dynamo uses the aws-sdk aws credentials implementation, this fixes
the case where ENV doesnt contain the aws credentials, where instead, the
credentials are fetched from the IAM role in the autoscaling group.

steveburkett committed:
Attempt to fix Faraday timeout errors

By increasing read and open timeouts.

steveburkett committed:
Add a synchronous option to elasticsearch create/delete

jpslav commented on app/services/books/index.rb:
```diff
@@ -52,8 +58,22 @@ def exists?
+      until self.send(this_happens) || tries > 5  do
+        Rails.logger.debug("Waiting for #{name} 1 sec for #{name} to #{this_happens.to_s}, tries: #{tries}")
+        sleep(2)
+        tries += 1
```
Is 10 seconds a long enough timeout?  And if we do hit the timeout should we raise an exception?  Otherwise it'll just explode later (at least our code will)

steveburkett replied:
yeah, let's raise it to 1 minute and raise an exception (a minute would justify an exception)

jpslav reviewed (changes requested):
The expect `Books::Index` spec should exercise `with_wait: true` so that the waiting code has some spec mileage.

steveburkett committed:
Code review fixes

make with_wait default to true (normal usage)
raise exception after a minute of waiting
re-record tests for index because of default use of with_wait

jpslav commented on app/services/books/indexing_strategies/i1/strategy.rb:
```diff
@@ -31,6 +31,12 @@ def index(book:, index_name:)
+    def total_number_of_elements_to_index(book:)
+      num_total = 0
+      book.root_book_part.pages.each{|page| num_total += page.elements(element_classes: DESIRED_ELEMENTS_TO_DOCUMENTS.keys).count }
+      num_total
```
This works, can also do `.map{...}.sum`.

Are we worried at all about parsing each page twice? Once here, once during indexing.

Can you also check me to see that pages and root book part is memoized (so we aren't actually downloading the book twice)

steveburkett replied:
def pages
      @pages ||= parts.map do |part|
        part.is_a?(Page) ? part : part.pages
      end.flatten
    end

jpslav commented on app/services/books/index.rb:
```diff
@@ -52,8 +63,27 @@ def exists?
+
+      if tries >= WAIT_UNTIL_MAX_TRIES
+        raise IndexResourceNotReadyError.new("Books::Index. #{name}:#{this_happens.to_s} failed after #{tries} tries.")
+      end
```
This could be at the end of the `until` loop above (changing `>=` to `>` probably which would let you remove the `|| tries > WAIT_UNTIL_MAX_TRIES` there.

jpslav commented on spec/cassettes/Books_Index/_name/derives_the_index_name.yml:
```diff
@@ -18,6 +18,11 @@ http_interactions:
+      - 299 Elasticsearch-6.7.2-56c6e48 "[types removal] The parameter include_type_name
+        should be explicitly specified in get indices requests to prepare for 7.0.
+        In 7.0 include_type_name will default to 'false', which means responses will
+        omit the type name in mapping definitions."
```
Good to know for the future :-)

jpslav commented on spec/cassettes/Books_Index/_delete/deletes_the_index.yml:
```diff
@@ -80,7 +95,39 @@ http_interactions:
+      - ''
+  response:
+    status:
+      code: 200
```
Nice to see the check for deletion failing.
